### PR TITLE
Fixes #1126: enable built-in expression in $orderby

### LINKED ISSUE #1126: Support tolower() in $orderby option
**Assemblies affected**
Which assemblies and versions are known to be affected e.g. ASP.NET Core OData 8.x
`Microsoft.AspNetCore.OData 8.2.3`

**Describe the bug**
I am not able to use a `tolower(text_field_in_model)` in an `$orderby` clause.

**Reproduce steps**
Hit a request against a model that has a search string:
```
http://localhost/api/students?$orderby=tolower(name)
```

**Data Model**
```
public class Student
{
    [Key]
    public string Id {get; set;}
    public string Name {get; set;}
}
```

**EDM (CSDL) Model**
```
<EntityType Name="Student">
    <Key>
        <PropertyRef Name="id" />
    </Key>
    <Property Name="id" Type="Edm.String" Nullable="false" />
    <Property Name="name" Type="Edm.String" Nullable="false" />
</EntityType>
```

**Request/Response**
http://localhost/api/students?$orderby=tolower(name)

**Expected behavior**
Results should be ordered by name (the property used in the $orderby clause)

**Actual behavior**
```
The query specified in the URI is not valid. Only ordering by properties is supported for non-primitive collections. Expressions are not supported.
```

**Screenshots**
NA

**Additional context**
NA

I just checked the code and it appears only `SingleValuePropertyAccessNode` is supported, not `SingleValueFunctionCallNode`.

P.S.: This is a valid OData construct - https://services.odata.org/V4/TripPinService/People?$orderby=tolower(FirstName)%20desc,UserName

### PR DESCRIPTION
**Pull request by xuzhg**
Fixes #1126

Support odata query expression in $orderby. For example:

$orderby=tolower(name)
$orderby=year(Birthday)
....

### CODE REVIEW

File: src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs [hunk 1]
```diff
@@ -32,59 +34,607 @@ public virtual void Validate(OrderByQueryOption orderByOption, ODataValidationSe
...
 
-            bool enableOrderBy = orderByOption.Context.DefaultQueryConfigurations.EnableOrderBy;
-            OrderByModelLimitationsValidator validator = new OrderByModelLimitationsValidator(orderByOption.Context, enableOrderBy);
-            bool explicitAllowedProperties = validationSettings.AllowedOrderByProperties.Count > 0;
+        /// <summary>
+        /// Override this method to validate property accessors.
+        /// </summary>
+        /// <param name="propertyAccessNode">The single value property access node.</param>
+        /// <param name="validatorContext">The validation context.</param>
+        protected virtual void ValidateSingleValuePropertyAccessNode(SingleValuePropertyAccessNode propertyAccessNode, OrderByValidatorContext validatorContext)
```
Comment (habbes) [inline src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs:286]: Did you add any tests on validation errors?

File: src/Microsoft.AspNetCore.OData/Query/Nodes/OrderByClauseNode.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+//------------------------------------------------------------------------------
+
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.AspNetCore.OData.Query
+{
+    /// <summary>
+    /// Represents the order by expression in the $orderby clause.
+    /// Use this to represent other $orderby except 'Property,OpenProperty,$count, $it' orderBy expression.
+    /// Again, in the next major release, we don't need this class.
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.AspNetCore.OData/Query/Nodes/OrderByClauseNode.cs:15]: #resolved

File: src/Microsoft.AspNetCore.OData/Query/Nodes/OrderByNode.cs [hunk 3]
```diff
@@ -14,6 +14,7 @@ namespace Microsoft.AspNetCore.OData.Query
 {
     /// <summary>
     /// Represents a single order by expression in the $orderby clause.
+    /// saxu: Why do we need this class and its derived type? only fetch the PropertyPath? In the next major release, we can consider to remove all of these.
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.AspNetCore.OData/Query/Nodes/OrderByNode.cs:17]: #resolved

File: src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs (hunk 1 above)
Comment (xuzhg) [reply src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs:286]: We already have the orderby validator tests in the UT: OrderByQueryValidatorTest.cs

File: src/Microsoft.AspNetCore.OData/Query/Validator/QueryValidatorHelpers.cs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                case ClrCanonicalFunctions.FloorFunctionName:
+                    result = AllowedFunctions.Floor;
+                    break;
+                case ClrCanonicalFunctions.HourFunctionName:
+                    result = AllowedFunctions.Hour;
+                    break;
+                case ClrCanonicalFunctions.IndexofFunctionName:
+                    result = AllowedFunctions.IndexOf;
+                    break;
+                case "isof":
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.AspNetCore.OData/Query/Validator/QueryValidatorHelpers.cs:84]: I don't know why. But, I prefer to use "isof" directly in the code.

[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.AspNetCore.OData/Query/Validator/QueryValidatorHelpers.cs:84]: #resolved.

**Review by habbes — APPROVED** [review]

File: src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs [hunk 5]
```diff
@@ -32,59 +34,607 @@ public virtual void Validate(OrderByQueryOption orderByOption, ODataValidationSe
...
+            }
+
+            if (countNode?.SearchClause != null)
+            {
+                ValidateQueryNode(countNode.SearchClause.Expression, validatorContext);
+            }
+        }
+
+        /// <summary>
+        /// override this method to restrict the binary operators inside the $orderby query.
```
Comment (gathogojr) [inline src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs:252]: nit:
```suggestion
        /// Override this method to restrict the binary operators inside the $orderby query.
```

File: src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs [hunk 6]
```diff
@@ -32,59 +34,607 @@ public virtual void Validate(OrderByQueryOption orderByOption, ODataValidationSe
...
+        protected virtual void ValidateSingleComplexNode(SingleComplexNode singleComplexNode, OrderByValidatorContext validatorContext)
+        {
+            if (singleComplexNode == null || validatorContext == null)
+            {
+                return;
+            }
+
+            // Check whether the property is sortable.
+            if (validatorContext.ValidationSettings.AllowedOrderByProperties.Count == 0)
+            {
```
Comment (gathogojr) [inline src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs:366]: Why do we execute this block when the expression `validatorContext.ValidationSettings.AllowedOrderByProperties.Count == 0`? I'd think that when the `AllowedOrderByProperties` collection is empty there's no need to call `EdmHelpers.IsNotSortable`... Or is my understanding wrong?

File: test/Microsoft.AspNetCore.OData.E2E.Tests/ODataOrderByTest/OrderByAdvancedTest.cs [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+
+            string payloadBody = await response.Content.ReadAsStringAsync();
+
+            // Origin is : 9, 8, 7, 1, 2
+            Assert.Equal("{\"Grades\":[1,2,7,8,9]}", payloadBody);
+        }
+
+
+        [Fact]
+        public async Task DollarOrderBy_UsingDollarThis_OnCollection()
```
Comment (gathogojr) [inline test/Microsoft.AspNetCore.OData.E2E.Tests/ODataOrderByTest/OrderByAdvancedTest.cs:229]:
```suggestion
        public async Task DollarOrderBy_UsingDollarIt_OnCollection()
```

**Comment by gathogojr** [comment]
Curious there'd be any value in adding E2E tests to verify the behaviour against EF query provider.

**Review by gathogojr — APPROVED** [review]
Left a few comments, otherwise LGTM

File: src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs (hunk 5 above)
Comment (xuzhg) [reply src/Microsoft.AspNetCore.OData/Query/Validator/OrderByQueryValidator.cs:252]: #Resolved.

**Pull request merged**